### PR TITLE
new: add possibility to retrieve details about rules with filter_details_visitor

### PR DESCRIPTION
**Pull request by loresuso**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind release

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area tests

> /area proposals

> /area CI

**What this PR does / why we need it**:
This PR introduces a specialized visitor that can be used to construct a normalized form of a rule, containing all the macros, lists, operators, and fields it uses in JSON format. The functionality is embedded in the `-l`/`-L` flags and can be triggered when the output is set to JSON. 
This functionality can be used later in CI tools for rules to e.g understand how to automatically handle and increase rules version.

Example:
```
$ sudo ./userspace/falco/falco -c ../falco.yaml -l "Launch Privileged Container" -o "json_output=true" | jq 
Fri May 12 10:23:13 2023: Falco version: 0.34.1-194+f5c7574 (x86_64)
Fri May 12 10:23:13 2023: Falco initialized with configuration file: ../falco.yaml
Fri May 12 10:23:13 2023: Loading rules from file /etc/falco/falco_rules.yaml
Fri May 12 10:23:13 2023: Loading rules from file /etc/falco/falco_rules.local.yaml
{
  "conditionFields": [
    "evt.num",
    "container.id",
    "container.privileged",
    "container.image.repository",
    "proc.vpid",
    "evt.dir",
    "evt.type"
  ],
  "description": "Detect the initial process started in a privileged container. Exceptions are made for known trusted images.",
  "exceptionFields": [],
  "lists": [
    "redhat_io_images_privileged",
    "falco_privileged_images",
    "trusted_images"
  ],
  "macros": [
    "redhat_image",
    "aws_eks_core_images",
    "user_trusted_containers",
    "allowed_openshift_registry_root",
    "openshift_image",
    "never_true",
    "falco_privileged_containers",
    "user_privileged_containers",
    "container",
    "spawned_process",
    "allowed_aws_ecr_registry_root_for_eks",
    "container_started"
  ],
  "name": "Launch Privileged Container",
  "operators": [
    "endswith",
    "startswith",
    "in",
    "!=",
    "="
  ],
  "outputFields": [
    "user.name",
    "user.loginuid",
    "proc.cmdline",
    "proc.pid",
    "container.name",
    "container.id",
    "container.image.repository",
    "container.image.tag"
  ],
  "priority": "Informational"
}
```

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
update: get rules details with `-l` or `-L` flags when json output format is specified
```

### CODE REVIEW

**Comment by loresuso** [comment]
cc @jasondellaluce

**Comment by FedeDP** [comment]
This is super cool and highly requested! Thank you @loresuso !

**Comment by jasondellaluce** [comment]
Links to https://github.com/falcosecurity/falco/issues/2372 and potentially fixes it, if we just add the information about the `evt.type`s matched by each rule.

**Comment by loresuso** [comment]
Last two commits address the following:
- retrieving information about enabled rules
- information about lists and macros when the global -L option is specified
This will allow to check for instance:
- if a rule used to be disabled and then enabled, patch number of version of the rule must be increased
- adding, removing elements from rules, same as above

Currently working on adding `evt.type` information

File: userspace/engine/rule_loader_compiler.cpp [hunk 1]
```diff
@@ -386,7 +386,11 @@ void rule_loader::compiler::compile_rule_infos(
 	std::string err, condition;
 	std::set<falco::load_result::load_result::warning_code> warn_codes;
 	filter_warning_resolver warn_resolver;
-	for (auto &r : col.rules())
+
+	// note: cast away the const qualifier in the for loop
+	// this is needed because we want to store information about evttypes
+	// used by any rules, which might come in handy when describing rules.
+	for (auto &r : const_cast<indexed_vector<rule_info>&>(col.rules()))
```
Comment (loresuso) [inline userspace/engine/rule_loader_compiler.cpp:393]: Added information about evttypes as well. However, there is a little bit of a hack.
The collector returns rule infos in a const indexed vector. The only part of code that has information about evt codes (or the complete ast with all macros resolved) is this one. So we may want to add this information to rule_info, but we can't given the const qualifier. For this reason we a couple of options:
- stop using const. This is not considerable as a good option, since const correctness is correctly throughout this part of code, and I don't like remove it just for this use case. Removing it here would touch a chain of const object, so it would have effect also in other parts of the code
- use `const_cast` to cast away the const qualifier only in this case, and to not break the clean interface we have with these classes. This is the cleanest solution I found to this problem, happy to discuss alternatives if any!

File: userspace/falco/app/actions/load_rules_files.cpp [hunk 2]
```diff
@@ -126,13 +126,13 @@ falco::app::run_result falco::app::actions::load_rules_files(falco::app::state&
 
 	if (s.options.describe_all_rules)
```
Comment (Andreagit97) [inline userspace/falco/app/actions/load_rules_files.cpp:127]: I would initialize both `describe_all_rules` and `describe_rule` variables here

```cpp
options::options()
	: event_buffer_format(sinsp_evt::PF_NORMAL),
	  gvisor_config(""),	
	  list_fields(false),
	  list_plugins(false),
	  list_syscall_events(false),
	  markdown(false),
	  modern_bpf(false)
{
}
```

**Review by Andreagit97 — COMMENTED** [review]
Left some minor comments, LGTM!

File: userspace/engine/filter_details_resolver.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	std::unordered_set<std::string> operators;
+	std::unordered_set<std::string> lists;
+
+	void reset();
+};
+
+/*!
+	\brief Helper class for getting details about rules' filters.
+*/
+class filter_details_resolver
```
Comment (Andreagit97) [inline userspace/engine/filter_details_resolver.h:42]: maybe we could add some tests to this class

File: userspace/engine/falco_engine.cpp [hunk 4]
```diff
@@ -427,25 +431,204 @@ std::size_t falco_engine::add_source(const std::string &source,
...
+
+	Json::FastWriter writer;
+	std::string json_str;
+	filter_details details;
+	for(const auto &m : m_rule_collector.macros())
+	{
+		// Assumption: no exception because rules have already been loaded.
+		auto cond_ast = libsinsp::filter::parser(m.cond).parse();
+		std::shared_ptr<libsinsp::filter::ast::expr> cond_ast_ptr = std::move(cond_ast);
+		details.known_macros[m.name] = cond_ast_ptr;
```
[the remark this answers is not in the document]
Comment (loresuso) [reply userspace/engine/falco_engine.cpp]: Slightly reworked the PR, not needed anymore :)

File: userspace/engine/filter_details_resolver.h (hunk 3 above)
Comment (loresuso) [reply userspace/engine/filter_details_resolver.h:42]: Agreed, I am adding this

File: userspace/falco/app/actions/load_rules_files.cpp (hunk 2 above)
Comment (loresuso) [reply userspace/falco/app/actions/load_rules_files.cpp:127]: They should already have their default I guess in options.cpp

File: userspace/engine/falco_engine.cpp [hunk 5]
```diff
@@ -427,25 +431,204 @@ std::size_t falco_engine::add_source(const std::string &source,
 	return m_sources.insert(src, source);
 }
 
-void falco_engine::describe_rule(std::string *rule) const
+void falco_engine::describe_rule(std::string *rule, bool json) const
```
[the remark this answers is not in the document]
Comment (loresuso) [reply userspace/engine/falco_engine.cpp:434]: Tried to address this, but figured that is better like it is already. Normally I agree with you, but here we have different behaviors when listing all rules in json or all rules in the old format, that is with the first one we have also to append all macros and lists information. However, with last commits duplication of code is reduced to minimum :)

File: userspace/engine/falco_engine.cpp [hunk 6]
```diff
@@ -427,27 +431,288 @@ std::size_t falco_engine::add_source(const std::string &source,
...
+	{
+		if(m_rule_collector.lists().at(i) != nullptr)
+		{
+			lists.append(i);
+			continue;
+		}
+		items.append(i);
+	}
+
+	list["details"]["items"] = items;
```
Comment (jasondellaluce) [inline userspace/engine/falco_engine.cpp:642]: Shouldn't `items` be inside `info` alongside `name`?

File: userspace/engine/falco_engine.cpp [hunk 7]
```diff
@@ -427,27 +431,288 @@ std::size_t falco_engine::add_source(const std::string &source,
...
+
+	// Parse rule condition and build the AST
+	// Assumption: no exception because rules have already been loaded.
+	auto ast = libsinsp::filter::parser(ri.cond).parse();
+	Json::Value json_details;
+	get_json_details(ast.get(), json_details);
+	rule["details"] = json_details;
+
+	// Get fields from output string
+	auto insp = new sinsp;
```
Comment (jasondellaluce) [inline userspace/engine/falco_engine.cpp:554]: I'd suggest passing an inspector as a parameter to the function, so that we don't have to re-alloc a new one for every rule. Moreover, i suggest wrapping it into a unique_ptr to make sure it is properly deallocated without having to deal with manual deletion. Alternatively, you could do a stack allocation and just pass the reference.

File: userspace/engine/falco_engine.cpp [hunk 8]
```diff
@@ -427,27 +431,288 @@ std::size_t falco_engine::add_source(const std::string &source,
...
+
+	// Get operators from exceptions
+	Json::Value exception_operators = Json::arrayValue;
+	for(const auto &e : ri.exceptions)
+	{
+		if(e.comps.is_list)
+		{
+			for(const auto& c : e.comps.items)
+			{
+				exception_operators.append(c.item);
```
Comment (jasondellaluce) [inline userspace/engine/falco_engine.cpp:582]: I truly don't remember if `c` can be a list here. Just to be sure, I would do a `c.is_list` check here.

File: userspace/engine/falco_engine.cpp [hunk 9]
```diff
@@ -427,27 +431,288 @@ std::size_t falco_engine::add_source(const std::string &source,
...
+	sinsp_evt_formatter fmt(insp, r.output);
+	std::vector<std::string> out_fields;
+	fmt.get_field_names(out_fields);
+	delete insp;
+	Json::Value outputFields = Json::arrayValue;
+	for(const auto &of : out_fields)
+	{
+		outputFields.append(of);
+	}
+	rule["details"]["outputFields"] = outputFields;
```
Comment (jasondellaluce) [inline userspace/engine/falco_engine.cpp:564]: Question here and for the ones below: should we use camel case for the json fields? I personally like it, but let's try to be consistent with the other Falco json outputs. If we usually adopt snake_case, then I'd suggest using it here too.

File: userspace/engine/falco_engine.cpp [hunk 10]
```diff
@@ -427,27 +431,288 @@ std::size_t falco_engine::add_source(const std::string &source,
...
+		{
+			auto evt_info = libsinsp::events::info(e);
+			auto res = evts.insert(std::string(evt_info->name));
+			if(res.second)
+			{
+				output.append(evt_info->name);
+			}
+		}
+	}
+}
```
Comment (jasondellaluce) [inline userspace/engine/falco_engine.cpp:713]:
```suggestion
	auto evtcodes = libsinsp::filter::ast::ppm_event_codes(ast);
	if(evtcodes.size() != libsinsp::events::all_event_set().size())
	{
		auto syscodes = libsinsp::filter::ast::ppm_sc_codes(ast);
		for (const auto& n : libsinsp::events::sc_set_to_event_names(syscodes))
		{
			output.append(n);
		}
		for (const auto& n : libsinsp::events::event_set_to_names(evtcodes, false))
		{
			output.append(n);
		}
	}
}
```

I know it's not intuitive... but I think that's what we want (this is how evt.type works under the hood):
- names from the event table for all non-generic events
- names from the syscall table for all generic events

File: userspace/engine/filter_details_resolver.cpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+{
+
+}
+
+void filter_details_resolver::visitor::visit(ast::binary_check_expr* e)
+{
+	m_details.fields.insert(e->field);
+	m_details.operators.insert(e->op);
+	
+	auto list = dynamic_cast<ast::list_expr*>(e->value.get());
```
[the remark this answers is not in the document]
Comment (loresuso) [reply userspace/engine/filter_details_resolver.cpp]: Thanks for this comment! I tried to address it in the last commit, mind to take a look again?

File: userspace/engine/filter_details_resolver.cpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+void filter_details_resolver::visitor::visit(ast::unary_check_expr* e)
+{
+	m_details.fields.insert(e->field);
+	m_details.operators.insert(e->op);
+}
+
+void filter_details_resolver::visitor::visit(ast::value_expr* e)
+{
```
[the remark this answers is not in the document]
Comment (loresuso) [reply userspace/engine/filter_details_resolver.cpp:88]: Same as above

File: userspace/engine/falco_engine.cpp (hunk 6 above)
Comment (loresuso) [reply userspace/engine/falco_engine.cpp]: Yes, agreed!

File: userspace/engine/falco_engine.cpp (hunk 7 above)
Comment (loresuso) [reply userspace/engine/falco_engine.cpp]: Agreed as well and addressed

File: userspace/engine/falco_engine.cpp (hunk 10 above)
Comment (loresuso) [reply userspace/engine/falco_engine.cpp]: Thank you, I totally missed this. Added this suggestion plus the merging of the two sets so that we don't have duplicates

File: userspace/engine/falco_engine.cpp (hunk 8 above)
Comment (loresuso) [reply userspace/engine/falco_engine.cpp:582]: In case it can be a list, what would you do? We should use a recursive approach or just check that it is not a list?
I would do something like:
```
if(!c.is_list)
  exception_operators.append(c.item)
```
but that alone wouldn't address the case of a list

File: userspace/engine/filter_details_resolver.cpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+void filter_details_resolver::visitor::visit(ast::and_expr* e)
+{
+	for(size_t i = 0; i < e->children.size(); i++)
+	{
+		m_expect_macro = true;
+		e->children[i]->accept(this);
+	}
+}
```
Comment (jasondellaluce) [inline userspace/engine/filter_details_resolver.cpp:42]:
```suggestion
		m_expect_macro = false;
	}
}
```

File: userspace/engine/filter_details_resolver.cpp [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+void filter_details_resolver::visitor::visit(ast::or_expr* e)
+{
+	for(size_t i = 0; i < e->children.size(); i++)
+	{
+		m_expect_macro = true;
+		e->children[i]->accept(this);
+	}
+}
```
Comment (jasondellaluce) [inline userspace/engine/filter_details_resolver.cpp:51]:
```suggestion
		m_expect_macro = false;
	}
}
```

File: userspace/engine/filter_details_resolver.cpp [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+void filter_details_resolver::visitor::visit(ast::not_expr* e)
+{
+	e->child->accept(this);
+}
+
+void filter_details_resolver::visitor::visit(ast::list_expr* e)
+{
+	if(m_expect_list)
+	{
+		m_expect_list = false;
```
Comment (jasondellaluce) [inline userspace/engine/filter_details_resolver.cpp:62]:
```suggestion
```

File: userspace/engine/filter_details_resolver.cpp [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+		}
+	}
+}
+
+void filter_details_resolver::visitor::visit(ast::binary_check_expr* e)
+{
+	m_details.fields.insert(e->field);
+	m_details.operators.insert(e->op);
+	m_expect_list = true;
+	e->value->accept(this);
```
Comment (jasondellaluce) [inline userspace/engine/filter_details_resolver.cpp:78]:
```suggestion
	m_expect_list = true;
	e->value->accept(this);
	m_expect_list = false;
```

File: userspace/engine/filter_details_resolver.cpp [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+{
+	m_details.fields.insert(e->field);
+	m_details.operators.insert(e->op);
+}
+
+void filter_details_resolver::visitor::visit(ast::value_expr* e)
+{
+	if(m_expect_macro)
+	{
+		m_expect_macro = false;
```
Comment (jasondellaluce) [inline userspace/engine/filter_details_resolver.cpp:91]:
```suggestion
```

File: userspace/engine/filter_details_resolver.cpp [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+			{
+				m_details.lists.insert(item);
+			}
+		}
+	}
+}
+
+void filter_details_resolver::visitor::visit(ast::binary_check_expr* e)
+{
+	m_details.fields.insert(e->field);
```
Comment (jasondellaluce) [inline userspace/engine/filter_details_resolver.cpp:75]:
```suggestion
	m_expect_macro = false;
	m_details.fields.insert(e->field);
```

File: userspace/engine/filter_details_resolver.cpp [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+	m_details.operators.insert(e->op);
+	m_expect_list = true;
+	e->value->accept(this);
+}
+
+void filter_details_resolver::visitor::visit(ast::unary_check_expr* e)
+{
+	m_details.fields.insert(e->field);
```
Comment (jasondellaluce) [inline userspace/engine/filter_details_resolver.cpp:83]:
```suggestion
	m_expect_macro = false;
	m_details.fields.insert(e->field);
```

File: userspace/engine/falco_engine.cpp (hunk 8 above)
Comment (jasondellaluce) [reply userspace/engine/falco_engine.cpp:582]: I think you could just do another level of `if(!c.is_list) { ... } else { ...}`. Recursion would be the safest option but I don't think exceptions go beyond 2 levels of lists.

**Review by jasondellaluce — COMMENTED** [review]
Looks good Lore:
- Left you a simple comment on that thing about exceptions and list of operators
- Left you suggestions to cover corner cases in the macro/list visitor search
- Once all's done, we'll need to rebase on top of https://github.com/falcosecurity/falco/pull/2552, that should bump the libs and hopefully fix the broken tests

**Comment by loresuso** [comment]
Everything's green, I think we are there now!

**Review by Andreagit97 — APPROVED** [review]
/approve
